File: apps/go/internal/chapter14/spending.go
package chapter14

import (
	"fmt"
	"slices"

	"github.com/k2works/getting-started-machine-learning/apps/go/internal/chapter02"
	"github.com/k2works/getting-started-machine-learning/apps/go/internal/chapter09"
)

// Categories は区分を表す番号で、支出額ではない列。
var Categories = []string{"Channel", "Region"}

// ClusterSummary は 1 つのクラスタの特徴。
type ClusterSummary struct {
	// Cluster はクラスタ番号
	Cluster int
	// Count は所属する件数
	Count int
	// Columns は Means の並びに対応する列名
	Columns []string
	// Means は元の単位での列ごとの平均
	Means []float64
}

// LoadSpending は Channel と Region を除いた支出額の列を読み込む。欠損値があればエラーを返す。
func LoadSpending(csvFile string) ([]chapter02.Features, error) {
	table, err := chapter02.LoadTable(csvFile)
	if err != nil {
		return nil, err
	}

	columns := make([]string, 0, len(table.Columns))

	for _, column := range table.Columns {
		if !slices.Contains(Categories, column) {
			columns = append(columns, column)
		}
	}

	x := make([]chapter02.Features, 0, len(table.Rows))

	for i, row := range table.Rows {
		values := make([]float64, len(columns))

		for j, column := range columns {
			value, ok, err := row.Number(column)
			if err != nil {
				return nil, err
			}

			if !ok {
				return nil, fmt.Errorf("%d 行目の %s が欠損値です", i+1, column)
			}

			values[j] = value
		}

		features, err := chapter02.NewFeatures(columns, values)
		if err != nil {
			return nil, err
		}

		x = append(x, features)
	}

	return x, nil
}

// StandardizeSpending は第 9 章の標準化（件数で割る標準偏差）で列ごとにそろえ、1 件を 1 つの点にする。
func StandardizeSpending(x []chapter02.Features) ([][]float64, error) {
	standardizer, err := chapter09.Fit(x)
	if err != nil {
		return nil, err
	}

	transformed, err := standardizer.Transform(x)
	if err != nil {
		return nil, err
	}

	points := make([][]float64, len(transformed))
	for i, features := range transformed {
		points[i] = features.Values
	}

	return points, nil
}

// SummarizeClusters はクラスタごとの件数と、元の単位での列ごとの平均を、件数の多い順に返す。
func SummarizeClusters(x []chapter02.Features, labels []int) ([]ClusterSummary, error) {
	if len(x) == 0 {
		return nil, fmt.Errorf("特徴量が 1 件もありません")
	}

	if len(x) != len(labels) {
		return nil, fmt.Errorf("特徴量とクラスタ番号の数が違います: %d と %d", len(x), len(labels))
	}

	columns := x[0].Columns
	clusters := make([]int, 0)
	members := make(map[int][]chapter02.Features)

	for i, label := range labels {
		if _, ok := members[label]; !ok {
			clusters = append(clusters, label)
		}

		members[label] = append(members[label], x[i])
	}

	slices.Sort(clusters)

	summaries := make([]ClusterSummary, 0, len(clusters))

	for _, cluster := range clusters {
		means, err := columnMeans(members[cluster], columns)
		if err != nil {
			return nil, err
		}

		summaries = append(summaries, ClusterSummary{
			Cluster: cluster,
			Count:   len(members[cluster]),
			Columns: slices.Clone(columns),
			Means:   means,
		})
	}

	// 件数が同じときにクラスタ番号の順が保たれるように、安定ソートで並べ替える
	slices.SortStableFunc(summaries, func(a, b ClusterSummary) int {
		return b.Count - a.Count
	})

	return summaries, nil
}

// columnMeans は列ごとの平均を求める。
func columnMeans(rows []chapter02.Features, columns []string) ([]float64, error) {
	means := make([]float64, len(columns))

	for j, column := range columns {
		sum := 0.0

		for _, features := range rows {
			value, err := features.Value(column)
			if err != nil {
				return nil, err
			}

			sum += value
		}

		means[j] = sum / float64(len(rows))
	}

	return means, nil
}

File: apps/ruby/lib/getting_started_ml/chapter14/spending.rb
# frozen_string_literal: true

module GettingStartedMl
  # 第 14 章の卸売業者の顧客ごとの支出額（Wholesale.csv）。
  module Chapter14
    # 区分を表す番号で、支出額ではない列。
    CATEGORIES = %w[Channel Region].freeze

    # 1 つのクラスタの特徴。クラスタ番号、所属する件数、列の順のままの列ごとの平均。
    ClusterSummary = Data.define(:cluster, :count, :means)

    module_function

    # Channel と Region を除いた支出額の列を読み込む。欠損値があれば失敗にする。
    def load_spending(csv_file)
      table = Chapter02::Table.load(csv_file)
      columns = table.columns - CATEGORIES

      table.rows.map do |row|
        values = columns.map { |column| row.number(column) || raise(ArgumentError, "欠損値があります: #{column}") }
        Chapter02::Features.new(columns:, values:)
      end
    end

    # 第 9 章の標準化（件数で割る標準偏差）で列ごとにそろえ、1 件を 1 行とする行列にする。
    def standardize_spending(x)
      Numo::DFloat.cast(Chapter09::Standardizer.fit(x).transform(x).map(&:values))
    end

    # クラスタごとの件数と、元の単位での列ごとの平均を、件数の多い順に並べる。件数が同じなら番号の順。
    def summarize_clusters(x, labels)
      summaries = x.zip(labels).group_by(&:last).map { |cluster, members| summarize(cluster, members.map(&:first)) }

      summaries.sort_by { |summary| [-summary.count, summary.cluster] }
    end

    # 1 つのクラスタに属する特徴量の並びを要約する。
    def summarize(cluster, members)
      means = members.map(&:values).transpose.map { |column| column.sum / column.size }

      ClusterSummary.new(cluster:, count: members.size, means:)
    end

    # クラスタの要約を、タブで区切った 1 行にする。平均は整数に丸める（Rust 版の {:.0} と同じく偶数丸め）。
    def format_summary(summary)
      [summary.cluster, summary.count, *summary.means.map { |mean| format("%.0f", mean) }].join("\t")
    end
  end
end
